src/openfermion/utils/_operator_utils.py

```diff
@@ -846,3 +847,86 @@ def normal_ordered(operator, hbar=1.):
+    in expectation estimation. Grouping into these tensor product basis
+    sets has been found to improve the efficiency of expectation estimation
+    significantly for some Hamiltonians in the context of
+    VQE (see section V(A) in the supplementary material of
+    https://arxiv.org/pdf/1704.05018v2.pdf). The more general problem
+    of grouping operators into commutitative groups is discussed in
+    section IV (B2) of https://arxiv.org/pdf/1509.04279v1.pdf. The
+    original input operator is the union of all output sub-operators,
+    and all sub-operators are disjoint (do not share any terms).
+
+    Args:
+        operator (QubitOperator): the operator that will be split into
+            sub-operators (tensor product basis sets).
+        seed (int): default None. Random seed used to initialize the
+            numpy.RandomState pseudo-random number generator.
+
+    Returns:
+        sub_operators (dict): a dictionary where each key defines a
+            tensor product basis, and each corresponding value is a
+            QubitOperator with terms that are all diagonal in
+            that basis.
+            **key** (tuple of tuples): Each key is a term, which defines
+                a tensor product basis. A term is a product of individual
+                factors; each factor is represented by a tuple of the form
+                (`index`, `action`), and these tuples are collected into a
+                larger tuple which represents the term as the product of
+                its factors. `action` is from the set {'X', 'Y', 'Z'} and
+                `index` is a non-negative integer corresponding to the
+                index of a qubit.
+            **value** (QubitOperator): A QubitOperator with terms that are
+                diagonal in the basis defined by the key it is stored in.
+
+    Raises:
+       TypeError: Operator of invalid type.
+    """
+
+    if not isinstance(operator, QubitOperator):
+        raise TypeError('Can only split QubitOperator into tensor product'
+                        ' basis sets. {} is not supported.'.format(
+                            type(operator).__name__))
+
+    sub_operators = {}
+    r = RandomState(seed)
+    for term, coefficient in operator.terms.items():
+        bases = list(sub_operators.keys())
+        r.shuffle(bases)
+
+        basis = _find_compatible_basis(term, bases)
+
+        if basis is None:
+            sub_operators[term] = QubitOperator(term, coefficient)
+        else:
+            additions = tuple(op for op in term if op not in basis)
+
+            if additions:
+                new_basis = tuple(sorted(basis + additions,
+                                         key=lambda factor: factor[0]))
+                sub_operators[new_basis] = sub_operators.pop(basis)
+                sub_operators[new_basis] += QubitOperator(term, coefficient)
+            else:
+                sub_operators[basis] += QubitOperator(term, coefficient)
```

Comment:
Get rid of the inner if-else clause by replacing the entire outer else clause with
```
sub_operator = sub_operators.pop(basis)
sub_operator += QubitOperator(term, coefficient)
additions = tuple(op for op in term if op not in basis)
basis = tuple(sorted(basis + additions, key=lambda factor: factor[0]))
sub_operators[basis] = sub_operator
```

Reply:
Sure, done